Call parse_links via a single call from Django middleware, not Jinja templates

richaagarwal opened the pull request:
## Overview
See GHE/CFGOV/enhanced-cms/issues/7 for context.  In summary, we have a lot of parts of the site that we don't call `parse_links` on, which means we don't add the external link redirect/icon/download link to links that should have them.  Because the set-up we currently have requires that `parse_links` is added to the correct place in the template and is called per Wagtail block, it involves a lot of mental and code overhead to ensure this is working across the board.  Instead, this PR centralizes the call via Django middleware, thus fixing most of the bugs outlined in the GHE issue above.  

## Notes 
- As a result of these changes, we are now parsing the entire page's html for links, instead of block by block, so I made some modifications to `parse_links` to ensure it only modifies the html that needs to, rather than returning the results from BeautifulSoup which adds a lot of unneeded markup. 
- We had several tests that were looking for specific html syntax in them, which has changed slightly as a result of these changes.  Nothing major, but we are adding a richtext filter on table cells now and for external links that are updated (that previously weren't), there are some minor markup changes. That's the main reason so many tests are updated. 
- `parse_links` used to remove style tags as part of its processing, but after discussion with @schaferjh and @jimmynotjim I've determined it's safe to remove that code and not really something we want to be/ still need to be doing.
- Functionally, things should work as before, with the main change being that external link icons, redirects, and download links should be applied consistently / correctly. Thus primarily this PR is a large bug fix.
- I know this PR is fairly large... it might be helpful to go through it commit by commit (I attempted to make each commit individually meaningful) to understand the changes as opposed to all at once.

## Testing
Will fill out soon

jimmynotjim commented:
This looks great to me, can't speak for the backend code though so you'll probably want a sign off from a BEWD as well.

hillaryj commented:
Code review looks good to me

chosak commented:
Testing this locally using the local Django server and a recent production dump, I can trigger a 500 (`AttributeError at / 'NavigableString' object has no attribute 'parent'`) on the home page ([http://localhost:8000](http://localhost:8000)). The particular error is caused by a tag with this content:

```html
<a class="" href="https://pueblo.gpo.gov/CFPBPubs/CFPBPubs.php">Order free brochures</a>
```

This tag exists in the "Order free brochures" link under "Practitioner Resources" in the megamenu.

This error seems to be caused by a space in the modified tag contents being rendered: if `tag.contents` [here](https://github.com/cfpb/cfgov-refresh/pull/4363/files#diff-b905b00a42016b4b740512bc7b0ae9f0R141) has a space as the second element like this, the error occurs on `str(tag)`:

```py
[<span class="a-link_text">Order free brochures</span>, u' ', <svg class="cf-icon-svg" ... ></svg>]
```

If I remove that second element (the empty space) it renders alright. I found [this](https://stackoverflow.com/questions/7591535/beautifulsoup-attributeerror-navigablestring-object-has-no-attribute-name) similar SO thread.

This seems related to [this line](https://github.com/cfpb/cfgov-refresh/pull/4363/files#diff-b905b00a42016b4b740512bc7b0ae9f0R138); I realize that [we did this before](https://github.com/cfpb/cfgov-refresh/pull/4363/files#diff-cfbd2caf2990c2080acb5d6bf65684f2L176), but perhaps doing the `str()` call is triggering something that wasn't happening before?

chosak commented on cfgov/core/utils.py:
```diff
@@ -1,23 +1,33 @@
+LINK_ICON_TEXT_CLASSES = os.environ.get(
+    'LINK_ICON_TEXT_CLASSES',
+    'a-link_text'
+)
```
Agree -- if we're not using these variables, let's remove reference to them.

richaagarwal replied:
@Scotchester Thanks - updated this in 39bb0c0f881999f3a9a7475185076d04dd0423c4

chosak commented:
After further debugging I was able to get this branch working. My version of BS4 was slightly out of date (4.3.2 instead of the current 4.5.1), and updating it fixed things.

Browsing around a local server, this seems to work nicely. I like this approach very much of using a middleware, to centralize these kinds of changes in one place instead of having them all over. There are some potential downsides - it'd be nice to understand what (if any) speed hit we're introducing by doing a full HTML page parse on ever page render. Additionally this calls [`parse_links`](https://github.com/cfpb/cfgov-refresh/pull/4363/files#diff-ff1364139d04d30c4db224e81c2946b0R20) against every single server response, including non-HTML things like JSON. In theory this would be okay because we wouldn't expect BS to match, but it'd be nice to filter those out somehow. Could we only run this logic on responses that have a `Content-Type` header matching [the default](https://docs.djangoproject.com/en/2.0/ref/settings/#default-content-type) (`text/html`)?

There is one piece of broken functionality that I notice: any Django views that return a [`StreamingHttpResponse`](https://docs.djangoproject.com/en/2.0/ref/request-response/#streaminghttpresponse-objects) instead of a regular `HttpResponse` will raise a 500 error, because those response types don't have a `content` attribute. The error will be something like

```
This FileResponse instance has no `content` attribute. Use `streaming_content` instead.
```

One example of views that use this are the Wagtail Document/Image links when running with local storage (e.g. not S3). To reproduce, using a local server try to upload an image at http://localhost:8000/admin/images/, then click through to your new image and try clicking on the filename to download. You'll trigger the 500.

[Here's](https://github.com/django/django/blob/master/django/middleware/gzip.py#L31) some example Django middleware that checks for `response.streaming` and handles streaming responses differently. [Here's](https://github.com/cobrateam/django-htmlmin/blob/master/htmlmin/middleware.py#L63) another nice pattern in a third-party middleware that uses a `can_minify_response` method to encapsulate the logic to determine if a response should be processed.

richaagarwal commented:
@chosak @Scotchester Thanks for the thorough reviews!  Will address individual comments soon. 

@chosak Regarding the bug you found, would it make sense to add `/f` to the blacklist I have?  I'm not sure which other URL patterns we might want to look out for.

chosak commented:
@richaagarwal I don't think that's best addressed by adding that to the blacklist, as that relies on the particulars of how the Django `MEDIA_ROOT` is set. Handling streaming responses more generally would handle any views where other the site use streaming responses, either now or in the future. For example, I notice that the CCDB [uses streaming responses](https://github.com/cfpb/ccdb5-api/blob/fc03a4f75c3ec6637a33796f9604577fdb220e26/complaint_search/views.py#L143).

richaagarwal commented:
@chosak I handle non-html content in this commit - 0697d24b9d5421e1002cd8c9a7d0ded14bfa9087.  This fixes the bug you observed with images.  Is something like this what you had in mind?

richaagarwal commented on cfgov/core/utils.py:
```diff
@@ -98,38 +100,46 @@ def is_image_tag(tag):
+    icon = False
+    if not tag.attrs.get('class', None):
+        tag.attrs.update({'class': []})
+    if tag['href'].startswith('/external-site/?'):
```
@Scotchester There are pages with this hardcoded in it... e.g. if you look at this page http://localhost:8000/admin/pages/2982/edit/ (locally if you have a production dump, or on content) you'll see it in the content section.  I think this is the case on all legacy pages.  I'm not too familiar with the history of these pages, but it looks like they were part of some kind of Wordpress migration and at some point we wanted to add in the external link icons differently than are defined in the hardcoded URL, hence the condition here.

Scotchester replied:
I guess it may not be feasible/worth the effort to replace all those instances? The code in this PR would now add the redirect if it wasn't hard-coded, right?

richaagarwal replied:
Yeah I think that's a little out of scope (by the way I didn't introduce this hardcoding, it was already a part of the function).

richaagarwal commented on cfgov/core/utils.py:
```diff
@@ -98,38 +100,46 @@ def is_image_tag(tag):
+        # Appends the SVG icon
+        tag.contents.append(BeautifulSoup(svg_icon(icon), 'html.parser'))
+        return str(tag)
+    return None
```
@Scotchester Updated in 3114eee4c804132747a1f04e894b6b97edc0566a

richaagarwal commented on cfgov/v1/__init__.py:
```diff
@@ -90,10 +89,6 @@ def environment(**options):
 
 
-def parse_links(value):
-    return value
```
@Scotchester Ah, good point. I think we're okay though -- I did a search for `parse_links` on the public CFPB org (https://github.com/search?q=org%3Acfpb+parse_links&type=Code) and there is one reference outside of cfgov-refresh, in https://github.com/cfpb/cfgov-sheer-templates/blob/9d474b9c766285c251b71656ffe56ef654db11ff/_includes/organisms/expandable.html, but I don't think we use that repo?  In github enterprise, if you do GHE/search?q=parse_links&type=Code, you'll see all the results (which searches all the orgs and all the repos) are off a particular user's fork of cfgov-refresh.

Scotchester replied:
Yeah, cfgov-sheer-templates is no longer used.

chosak commented:
@richaagarwal yes, thanks. I would recommend moving that method (and, honestly, `parse_links` itself) into the middleware, to centralize all of the logic in one place. Now that we only call `parse_links` from the middleware, there's no need to have it elsewhere, and making those methods on the middleware class makes for better encapsulation and might make for easier debugging/modification in future.

richaagarwal commented on cfgov/core/tests/test_middleware.py:
```diff
@@ -0,0 +1,29 @@
+    @mock.patch('core.utils.parse_links')
+    def test_parse_links_does_not_get_called(self, mock_parse_links):
+        """Middleware does not invoke parse links when path is in blacklist"""
+        self.client.get('/foo/bar')
```
@chosak I started writing tests for this and am a little stuck.  If you run this as is, the first test fails with 
```
AssertionError: Expected call: parse_links('')
Not called
```
I couldn't quite figure out why because as I was stepping through it, it was indeed getting called.  Then I removed the second test altogether and the first test passed.  So then I went through line by line and noticed that if line 28 (in the second test) is commented out, the first test passes. This seems odd because in the docs, https://docs.djangoproject.com/en/2.0/topics/testing/tools/#django.test.SimpleTestCase.client, it means that the client gets recreated in each test.  Any thoughts??

chosak replied:
@richaagarwal if you change the patch to `@mock.patch('core.middleware.parse_links')`, it works. I think you have to patch the module path as encountered by the test, not necessarily where it is imported from in the end. See "Patching in the wrong place" [here](http://alexmarandon.com/articles/python_mock_gotchas/).

richaagarwal replied:
@chosak ah, thanks so much!  I think I've probably encountered that before but never realized that was what was going on.

richaagarwal commented:
@chosak I've added more test coverage and moved `parse_links` and `should_parse_links` into the middleware here - e3761cf0740ff1239e76038a4d31bc3c971676f5.  Decided to leave the other util methods as is for now.  If that looks OK I'd like to merge :)
